### ISSUE TITLE
GraphQL Fragments + Typename warning in console

### ISSUE THREAD

**Issue by jcjp** [Mike-Gibson/mock-apollo-client#24]
I have multiple warning messages when using a GraphQL query with fragments, here are some of the warning:

```javascript
  console.warn
    You're using fragments in your queries, but either don't have the addTypename:
      true option set in Apollo Client, or you are trying to write a fragment to the store without the __typename.
       Please turn on the addTypename option and include __typename when writing fragments so that Apollo Client
       can accurately match fragments.
```

```javascript
  console.warn
    Could not find __typename on Fragment  A {
     ...
    }
```

```javascript
  console.warn
    DEPRECATION WARNING: using fragments without __typename is unsupported behavior and will be removed in future versions of Apollo client. You should fix this and set addTypename to true now.
```

- I tried adding `__typename` to my mock result however it results into an error
- Setting the `addTypename` to `true` would result to an error of `Request handler not defined for query` which redirected me to this github issue: [I always get "Request handler not defined for query:..."](https://github.com/Mike-Gibson/mock-apollo-client/issues/15).
- I also tried adding the following on the mock client (see below), But still results into `Request handler not defined for query`

```javascript
cache: new InMemoryCache({
 fragmentMatcher: new IntrospectionFragmentMatcher({
  introspectionQueryResultData,
 })
})
```

Currently all test are passing as of the moment. I am using version `mock-apollo-client": "^0.5.0` . Since this is just warnings and in unit testing I could simply ignore them.

**Comment by Mike-Gibson** [comment]
Hi, sorry for the delayed reply.

Yeah, this has been an issue before (which you found) and I didn't have time to repro it find a solution. 

Any chance you could set up a repo with a minimal example of its failing? I'd like to have a look and see if it's something that can be fixed.

It seems to be a combination of the cache, query matching and mock response data which causes it. 

For now, yes I guess you can ignore the warnings if it's working. It's possible to write some code in the jest setup script to intercept console.warn calls and suppress them if they match a pattern - that might be a simple way to make your test output less noisy for now.

**Comment by jcjp** [comment]
Sorry I am still unable to create a sample repo. Hopefully this weekend.

**Comment by jcjp** [comment]
Hi Mike I created a sample repository here: [Nuxt Apollo Unit Test](https://github.com/jcjp/nuxt-apollo-unit-test) I am using Contenful to query GraphQL data it is currently on 10 days trial and my account will expire. If you know some free data that we could query using GraphQL let me know.

**Comment by Mike-Gibson** [comment]
Thanks, that's great. I'll hopefully have time this weekend to take a look.

As for GraphQL APIs for testing, I've not actually used any myself. A quick google seems to turn up a few. Interestingly, github has one: https://docs.github.com/en/graphql/overview/explorer
but requires auth. It has a queries including fragments ie.
```graphql
query { 
   repository(owner:"Mike-Gibson", name:"mock-apollo-client") {
    issueOrPullRequest(number: 1) {
      __typename
      ... on Issue {
        author {
          login
        }
      }
      ... on PullRequest {
        author {
          login
        }
      }
    }
  }
}
```

**Comment by Mike-Gibson** [comment]
I've just pushed a new version - `0.6.0-feature-fragments.0` which should allow fragments in queries.
I haven't updated the docs yet - I'd like to get your feedback on whether it works for your use case first.

Could you give it a go and let me know how you get on?

You'll need to create the mock apollo client and customise the cache to add the typename handling:
```typescript
mockClient = createMockClient({
  cache: new InMemoryCache({
    addTypename: true,
  }),
});
```

and then in the mock response, add the typename field, i.e.

```typescript
const response = { data: { user: { __typename: 'User', name: 'Bob' } } };
```

The query itself shouldn't need changing (so __typename doesn't need to be added)

**Comment by jcjp** [comment]
I upgraded my version to `0.6.0-feature-fragments.0` and I am encountering an error when adding the `cache` on the `createMockClient`.

```javascript
TS2741: Property 'writeData' is missing in type 'InMemoryCache' but required in type 'ApolloCache<NormalizedCacheObject>'.
    51 |     defaultClient = createMockClient({
    52 |       resolvers: {},
  > 53 |       cache: new InMemoryCache({
       |       ^^^^^
    54 |         addTypename: true
    55 |       })
    56 |     })
```

I imported the `InMemoryCache` from `@apollo/client`. When I tried to ignore the TypeScript checking and run the test it shows an error below:

```javascript
$ jest
 PASS  test/Logo.spec.js
 FAIL  pages/index.spec.ts
  ● Test suite failed to run

    pages/index.spec.ts:53:7 - error TS2741: Property 'writeData' is missing in type 'InMemoryCache' but required in type 'ApolloCache<NormalizedCacheObject>'.

    53       cache: new InMemoryCache({
             ~~~~~

      node_modules/apollo-cache/lib/cache.d.ts:22:5
        22     writeData<TData = any>({ id, data, }: Cache.WriteDataOptions<TData>): void;
               ~~~~~~~~~
        'writeData' is declared here.
```

**Comment by Mike-Gibson** [comment]
Based on the previous issue and the sample repo, I thought you were using v2 of Apollo - so I think the import should be 

```typescript
import { InMemoryCache } from 'apollo-cache-inmemory';
```

I've updated the test code in my fork of your sample - https://github.com/Mike-Gibson/nuxt-apollo-unit-test/commit/bea9ecc02c7f8dc043c0bf70a232401af3e89e23

**Comment by jcjp** [comment]
> 
> 
> Based on the previous issue and the sample repo, I thought you were using v2 of Apollo - so I think the import should be
> 
> ```ts
> import { InMemoryCache } from 'apollo-cache-inmemory';
> ```
> 
> I've updated the test code in my fork of your sample - [Mike-Gibson/nuxt-apollo-unit-test@bea9ecc](https://github.com/Mike-Gibson/nuxt-apollo-unit-test/commit/bea9ecc02c7f8dc043c0bf70a232401af3e89e23)

I updated my code, yes you are correct I am using v2 of Apollo and tried running it again it seems that the console logs are now gone! Here is the result of the unit test:

![Unit Test Result](https://user-images.githubusercontent.com/37642320/113096507-e21f1600-9227-11eb-9fd9-d385453ecad9.jpg)

Thanks I guess we can now conclude and close this ticket! Just additional information just like Mike said I need to add the `__typename: sampleTypeName` for every object in GraphQL. Here is a sample from mine:

```javascript
"itemCollectionCollection": {
 "items": [
  {
    "itemsCollection": {
      "items": [
        {
          "question": "Sample Question 1",
           "answer": "Sample Answer 1",
           "__typename": "Data"
        },
        {
          "question": "Sample Question 2",
           "answer": "Sample Answer 2",
           "__typename": "Data"
         }
       ],
     "__typename": "ItemCollectionItemsCollection"
     },
    "__typename": "ItemCollection"
    }
 ],
"__typename": "ItemCollectionCollection"
}
```

**Comment by Mike-Gibson** [comment]
That's great news. OK, so I will tidy up the changes, and I'll also update the README to include the info about fragments. Then I'll create a new 0.6 release (the one I gave you is technically a pre-release version, but feel free to use for now). I'll also take a look at supporting fragments in mock-apollo-client v1 (Apollo 3) as I think the same approach should work.

Thanks for your help @jcjp. I'll let you know when I've released v0.6 - hopefully by the weekend
